File: lwjgl/lesson02/Lesson02/Lesson02.java
/*
 *      This Code Was Created By Jeff Molofee 2000
 *      A HUGE Thanks To Fredric Echols For Cleaning Up
 *      And Optimizing The Base Code, Making It More Flexible!
 *      If You've Found This Code Useful, Please Let Me Know.
 *      Visit My Site At nehe.gamedev.net
 */

import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;
import org.lwjgl.opengl.GL11;
import org.lwjgl.opengl.glu.GLU;
import org.lwjgl.input.Keyboard;

/**
 * @author Mark Bernard
 * date:    16-Nov-2003
 *
 * Port of NeHe's Lesson 2 to LWJGL
 * Title: Your First Polygon
 * Uses version 0.8alpha of LWJGL http://www.lwjgl.org/
 *
 * Be sure that the LWJGL libraries are in your classpath
 *
 * Ported directly from the C++ version
 *
 * 2004-05-08: Updated to version 0.9alpha of LWJGL.
 *             Changed from all static to all instance objects.
  * 2004-09-21: Updated to version 0.92alpha of LWJGL.
*/
public class Lesson02 {
    private boolean done = false;
    private boolean fullscreen = false;
    private final String windowTitle = "NeHe's OpenGL Lesson 2 for LWJGL (Your First Polygon)";
    private boolean f1 = false;
    private DisplayMode displayMode;

    public static void main(String args[]) {
        boolean fullscreen = false;
        if(args.length>0) {
            if(args[0].equalsIgnoreCase("fullscreen")) {
                fullscreen = true;
            }
        }

        Lesson02 l2 = new Lesson02();
        l2.run(fullscreen);
    }
    /**
     * Launch point
     * @param fullscreen boolean value, set to true to run in fullscreen mode
     */
    public void run(boolean fullscreen) {
        this.fullscreen = fullscreen;
        try {
            init();
            while (!done) {
                mainloop();
                render();
                Display.update();
            }
            cleanup();
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
    }

    /**
     * Very basic for first lesson.  Just check for escape key or user
     * clicking to close the window.
     */
    private void mainloop() {
        if(Keyboard.isKeyDown(Keyboard.KEY_ESCAPE)) {       // Exit if Escape is pressed
            done = true;
        }
        if(Display.isCloseRequested()) {                     // Exit if window is closed
            done = true;
        }
        if(Keyboard.isKeyDown(Keyboard.KEY_F1) && !f1) {    // Is F1 Being Pressed?
            f1 = true;                                      // Tell Program F1 Is Being Held
            switchMode();                                   // Toggle Fullscreen / Windowed Mode
        }
        if(!Keyboard.isKeyDown(Keyboard.KEY_F1)) {          // Is F1 Being Pressed?
            f1 = false;
        }
    }

    private void switchMode() {
        fullscreen = !fullscreen;
        try {
            Display.setFullscreen(fullscreen);
        }
        catch(Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * For rendering all objects to the screen
     * @return boolean for success or not
     */
    private void render() {
        GL11.glClear(GL11.GL_COLOR_BUFFER_BIT | GL11.GL_DEPTH_BUFFER_BIT);          // Clear The Screen And The Depth Buffer
        GL11.glLoadIdentity();                          // Reset The Current Modelview Matrix
        GL11.glTranslatef(-1.5f,0.0f,-6.0f);                // Move Left 1.5 Units And Into The Screen 6.0
        GL11.glBegin(GL11.GL_TRIANGLES);                    // Drawing Using Triangles
            GL11.glVertex3f( 0.0f, 1.0f, 0.0f);         // Top
            GL11.glVertex3f(-1.0f,-1.0f, 0.0f);         // Bottom Left
            GL11.glVertex3f( 1.0f,-1.0f, 0.0f);         // Bottom Right
        GL11.glEnd();                                       // Finished Drawing The Triangle
        GL11.glTranslatef(3.0f,0.0f,0.0f);              // Move Right 3 Units
        GL11.glBegin(GL11.GL_QUADS);                        // Draw A Quad
            GL11.glVertex3f(-1.0f, 1.0f, 0.0f);         // Top Left
            GL11.glVertex3f( 1.0f, 1.0f, 0.0f);         // Top Right
            GL11.glVertex3f( 1.0f,-1.0f, 0.0f);         // Bottom Right
            GL11.glVertex3f(-1.0f,-1.0f, 0.0f);         // Bottom Left
        GL11.glEnd();                                       // Done Drawing The Quad
    }

    /**
     * Create a window for all display
     * @throws Exception
     */
    private void createWindow() throws Exception {
        Display.setFullscreen(fullscreen);
        DisplayMode d[] = Display.getAvailableDisplayModes();
        for (int i = 0; i < d.length; i++) {
            if (d[i].getWidth() == 640
                && d[i].getHeight() == 480
                && d[i].getBitsPerPixel() == 32) {
                displayMode = d[i];
                break;
            }
        }
        Display.setDisplayMode(displayMode);
        Display.setTitle(windowTitle);
        Display.create();
    }

    /**
     * Initialize the environment
     * @throws Exception
     */
    private void init() throws Exception {
        createWindow();

        initGL();
    }

    /**
     * Initialize OpenGL
     *
     */
    private void initGL() {
        GL11.glEnable(GL11.GL_TEXTURE_2D); // Enable Texture Mapping
        GL11.glShadeModel(GL11.GL_SMOOTH); // Enable Smooth Shading
        GL11.glClearColor(0.0f, 0.0f, 0.0f, 0.0f); // Black Background
        GL11.glClearDepth(1.0); // Depth Buffer Setup
        GL11.glEnable(GL11.GL_DEPTH_TEST); // Enables Depth Testing
        GL11.glDepthFunc(GL11.GL_LEQUAL); // The Type Of Depth Testing To Do

        GL11.glMatrixMode(GL11.GL_PROJECTION); // Select The Projection Matrix
        GL11.glLoadIdentity(); // Reset The Projection Matrix

        // Calculate The Aspect Ratio Of The Window
        GLU.gluPerspective(
          45.0f,
          (float) displayMode.getWidth() / (float) displayMode.getHeight(),
          0.1f,
          100.0f);
        GL11.glMatrixMode(GL11.GL_MODELVIEW); // Select The Modelview Matrix

        // Really Nice Perspective Calculations
        GL11.glHint(GL11.GL_PERSPECTIVE_CORRECTION_HINT, GL11.GL_NICEST);
    }

    /**
     * Cleanup all the resources.
     *
     */
    private void cleanup() {
        Display.destroy();
    }

}

File: ruby/lesson02.rb
# (c) 2004 Ben Goodspeed
# Lesson 2-Ruby, based on the tutorials at nehe.gamedev.net
#   based on Jeff Molofee (1999), Cora Hessey (2002)(perl port).
#
# Requires the ruby opengl bindings from: http://www2.giganet.net/~yoshi/
require "opengl"
require "glut"

# keycode we match against in the keyboard listener
ESCAPE = 27
$width , $height = 640, 480


initGL = Proc.new {
    # clear to black
    GL.ClearColor(0.0,0.0,0.0,0.0)

    GL.ClearDepth(1.0)
    GL.DepthFunc(GL::LESS);         

    # Enables depth testing with that type
    GL.Enable(GL::DEPTH_TEST);              
    
    # Enables smooth color shading
    GL.ShadeModel(GL::SMOOTH);      

    # Reset the projection matrix
    GL.MatrixMode(GL::PROJECTION);
    GL.LoadIdentity;

    # Calculate the aspect ratio of the Window
    GLU.Perspective(45.0, $width/$height, 0.1, 100.0);

    # Reset the modelview matrix
    GL.MatrixMode(GL::MODELVIEW);


}

display = Proc.new {
    # Clear the screen and the depth buffer
    GL.Clear(GL::COLOR_BUFFER_BIT | GL::DEPTH_BUFFER_BIT);  

    # Reset the view
    GL.LoadIdentity;

    # Move to the left 1.5 units and into the screen 6.0 units
    GL.Translate(-1.5, 0.0, -6.0); 
        
    # -- Draw a triangle --
    GL.Color(1.0,1.0,1.0)
    # Begin drawing a polygon
    GL.Begin(GL::POLYGON);
      GL.Vertex3f( 0.0, 1.0, 0.0);     # Top vertex
      GL.Vertex3f( 1.0, -1.0, 0.0);    # Bottom right vertex
      GL.Vertex3f(-1.0, -1.0, 0.0);    # Bottom left vertex
    # Done with the polygon
    GL.End;

    # Move 3 units to the right
    GL.Translate(3.0, 0.0, 0.0);

    # -- Draw a square (quadrilateral) --
    # Begin drawing a polygon (4 sided)
    GL.Begin(GL::QUADS);
      GL.Vertex3f(-1.0, 1.0, 0.0);       # Top Left vertex
      GL.Vertex3f( 1.0, 1.0, 0.0);       # Top Right vertex
      GL.Vertex3f( 1.0, -1.0, 0.0);      # Bottom Right vertex
      GL.Vertex3f(-1.0, -1.0, 0.0);      # Bottom Left  
    GL.End;                
    GL.Flush
    # Since this is double buffered, swap the buffers.
    # This will display what just got drawn.
    GLUT.SwapBuffers;
}

reshape = Proc.new { |w,h|
    h = 1 if h == 0
    $width, $height = w,h
    GL.Viewport(0, 0, $width, $height);              
    # Re-initialize the window (same lines from InitGL)
    GL.MatrixMode(GL::PROJECTION);
    GL.LoadIdentity;
    GLU.Perspective(45.0, $width/$height, 0.1, 100.0);
    GL.MatrixMode(GL::MODELVIEW);
}

keyboard = Proc.new { |key,x,y|
    case (key)
        when ESCAPE
        exit 0
        when 'f'[0]
        GLUT.ReshapeWindow(640,480)
    end
}




# Initialize glut & open a window
GLUT.Init
GLUT.InitDisplayMode(GLUT::DOUBLE | GLUT::RGB | GLUT::DEPTH)
GLUT.InitWindowSize($width, $height)
GLUT.CreateWindow($0)

# initialize opengl
initGL.call

# add callback functions for some event listeners
GLUT.ReshapeFunc(reshape)
GLUT.DisplayFunc(display)
GLUT.KeyboardFunc(keyboard)
GLUT.FullScreen
# enter the main idle loop
GLUT.MainLoop()

